### PR TITLE
Fix wide integer left shift + refactoring

### LINKED ISSUE #14677: (unreleased) debug assertion in some arithmetic operations on UInt256
`SELECT bitShiftLeft(toInt256(-2), number)`

`shift left for negative lhsbers is underfined!`

See https://clickhouse-test-reports.s3.yandex.net/14525/44726c37c3003ec8112dbf32efa350bd595163b3/fuzzer/report.html#fail1

### PR DESCRIPTION
**Pull request by 4ertus2**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix crash in `bitShiftLeft()` when called with negative big integer.

Fixes #14677 

It's also interesting if we really need `noexcept` for methods at all https://stackoverflow.com/questions/21463119/does-it-make-sense-to-declare-inline-functions-noexcept

### CODE REVIEW

File: base/common/wide_integer_impl.h [hunk 1]
```diff
@@ -275,23 +283,19 @@ struct wide_integer<Bits, Signed>::_impl
         return lhs;
     }
 
-    constexpr static wide_integer<Bits, signed> shift_left(const wide_integer<Bits, signed> & rhs, int n)
+    constexpr static integer<Bits, signed> shift_left(const integer<Bits, signed> & rhs, int n) noexcept
     {
-        // static_assert(is_negative(rhs), "shift left for negative lhsbers is underfined!");
-        if (is_negative(rhs))
-            throw std::runtime_error("shift left for negative lhsbers is underfined!");
```
Comment (4ertus2) [inline base/common/wide_integer_impl.h:282]: Crach was here when called from `operator << () noexcept`

Comment (akuzm) [reply base/common/wide_integer_impl.h:282]: By the way, what is the point of `noexcept` there? I always thought its only function was making the code slower by inserting try/catch + std::terminate()...

Comment (akuzm) [reply base/common/wide_integer_impl.h:282]: On an arithmetic operator especially, it must be costing us a lot.

Comment (4ertus2) [reply base/common/wide_integer_impl.h:282]: I remove it in this PR and make `shift_left`noexcept.

File: base/common/throwError.h [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+#include <Poco/Exception.h>
+
+template <typename T>
+inline void throwError(const T & err)
```
Comment (alexey-milovidov) [inline base/common/throwError.h:5]: Maybe add a comment about motivation?

Comment (4ertus2) [reply base/common/throwError.h:10]: done

File: base/common/wide_integer_impl.h (hunk 1 above)
Comment (4ertus2) [reply base/common/wide_integer_impl.h:282]: Current library logic is to have `noexcept` all the places its possible. Probably we should change it for better performance but not in this PR.

**Comment by 4ertus2** [comment]
ya chek if broken by other test:  01473_event_time_microseconds

**Comment by 4ertus2** [comment]
Stress test was OK before 'merge master'. Most probably It's not related to PR changes.

**Comment by 4ertus2** [comment]
Ya check it OK after 'merge master', other tests (except perf) were OK before.

**Comment by 4ertus2** [comment]
There's a crash in CacheDictionary in Stress test server.log https://clickhouse-test-reports.s3.yandex.net/14697/4d5615cb121712ea2ab497800bda295539b49cd8/stress_test_(memory)/clickhouse-server.log
```
2020.09.14 13:41:08.638417 [ 9598 ] {} <Fatal> BaseDaemon: ########################################
2020.09.14 13:41:08.641745 [ 9598 ] {} <Fatal> BaseDaemon: (version 20.10.1.4639, build id: 92A8A8CEBAB7D5FE) (from thread 2146) (no query) Received signal Segmentation fault (11)
2020.09.14 13:41:08.643254 [ 9598 ] {} <Fatal> BaseDaemon: Address: 0x8008 Access: write. Address not mapped to object.
2020.09.14 13:41:08.643409 [ 9598 ] {} <Fatal> BaseDaemon: Stack trace: 0x272b2f3d 0x272b2ab3 0x272b1fbe 0x27318b1a 0x2727d687 0x2727b099 0x2728350c 0xf128135 0xf133acc 0xf1338fd 0xf1202ea 0xf12d11c 0x7f4287b8b669 0x7f4287aa22b3
2020.09.14 13:41:08.739189 [ 9598 ] {} <Fatal> BaseDaemon: 4. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/__hash_table:2412: std::__1::__hash_table<std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > >, std::__1::__unordered_map_hasher<unsigned long, std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > >, std::__1::hash<unsigned long>, true>, std::__1::__unordered_map_equal<unsigned long, std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > >, std::__1::equal_to<unsigned long>, true>, std::__1::allocator<std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > > > >::__rehash(unsigned long) @ 0x272b2f3d in /usr/bin/clickhouse
2020.09.14 13:41:08.754893 [ 9598 ] {} <Fatal> BaseDaemon: 5. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/__hash_table:0: std::__1::__hash_table<std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > >, std::__1::__unordered_map_hasher<unsigned long, std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > >, std::__1::hash<unsigned long>, true>, std::__1::__unordered_map_equal<unsigned long, std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > >, std::__1::equal_to<unsigned long>, true>, std::__1::allocator<std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > > > >::rehash(unsigned long) @ 0x272b2ab3 in /usr/bin/clickhouse
2020.09.14 13:41:08.778240 [ 9598 ] {} <Fatal> BaseDaemon: 6. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/__hash_table:0: std::__1::pair<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > >, void*>*>, bool> std::__1::__hash_table<std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > >, std::__1::__unordered_map_hasher<unsigned long, std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > >, std::__1::hash<unsigned long>, true>, std::__1::__unordered_map_equal<unsigned long, std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > >, std::__1::equal_to<unsigned long>, true>, std::__1::allocator<std::__1::__hash_value_type<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > > > >::__emplace_unique_key_args<unsigned long, std::__1::piecewise_construct_t const&, std::__1::tuple<unsigned long const&>, std::__1::tuple<> >(unsigned long const&, std::__1::piecewise_construct_t const&, std::__1::tuple<unsigned long const&>&&, std::__1::tuple<>&&) @ 0x272b1fbe in /usr/bin/clickhouse
2020.09.14 13:41:08.816225 [ 9598 ] {} <Fatal> BaseDaemon: 7. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/vector:1516: std::__1::__function::__func<void DB::CacheDictionary::getItemsNumberImpl<unsigned long, unsigned long, DB::CacheDictionary::getUInt64(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul> const&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul>&) const::$_3>(DB::CacheDictionary::Attribute&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul> const&, std::__1::conditional<IsDecimalNumber<unsigned long>, DB::DecimalPaddedPODArray<unsigned long>, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul> >::type&, DB::CacheDictionary::getUInt64(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul> const&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul>&) const::$_3&&) const::'lambda'(unsigned long, unsigned long), std::__1::allocator<void DB::CacheDictionary::getItemsNumberImpl<unsigned long, unsigned long, DB::CacheDictionary::getUInt64(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul> const&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul>&) const::$_3>(DB::CacheDictionary::Attribute&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul> const&, std::__1::conditional<IsDecimalNumber<unsigned long>, DB::DecimalPaddedPODArray<unsigned long>, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul> >::type&, DB::CacheDictionary::getUInt64(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul> const&, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul>&) const::$_3&&) const::'lambda'(unsigned long, unsigned long)>, void (unsigned long, unsigned long)>::operator()(unsigned long&&, unsigned long&&) @ 0x27318b1a in /usr/bin/clickhouse
2020.09.14 13:41:08.823691 [ 9598 ] {} <Fatal> BaseDaemon: 8. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:1825: DB::CacheDictionary::update(std::__1::shared_ptr<DB::CacheDictionary::UpdateUnit>&) const @ 0x2727d687 in /usr/bin/clickhouse
2020.09.14 13:41:08.827578 [ 9598 ] {} <Fatal> BaseDaemon: 9. /build/obj-x86_64-linux-gnu/../src/Dictionaries/CacheDictionary.cpp:0: DB::CacheDictionary::updateThreadFunction() @ 0x2727b099 in /usr/bin/clickhouse
2020.09.14 13:41:08.831634 [ 9598 ] {} <Fatal> BaseDaemon: 10. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:1714: std::__1::__function::__func<DB::CacheDictionary::CacheDictionary(DB::StorageID const&, DB::DictionaryStructure const&, std::__1::unique_ptr<DB::IDictionarySource, std::__1::default_delete<DB::IDictionarySource> >, DB::ExternalLoadableLifetime, unsigned long, unsigned long, bool, unsigned long, unsigned long, unsigned long, unsigned long)::$_4, std::__1::allocator<DB::CacheDictionary::CacheDictionary(DB::StorageID const&, DB::DictionaryStructure const&, std::__1::unique_ptr<DB::IDictionarySource, std::__1::default_delete<DB::IDictionarySource> >, DB::ExternalLoadableLifetime, unsigned long, unsigned long, bool, unsigned long, unsigned long, unsigned long, unsigned long)::$_4>, void ()>::operator()() @ 0x2728350c in /usr/bin/clickhouse
2020.09.14 13:41:08.834468 [ 9598 ] {} <Fatal> BaseDaemon: 11. /build/obj-x86_64-linux-gnu/../src/Common/CurrentMetrics.h:74: ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0xf128135 in /usr/bin/clickhouse
2020.09.14 13:41:08.835558 [ 9598 ] {} <Fatal> BaseDaemon: 12. /build/obj-x86_64-linux-gnu/../src/Common/ThreadPool.h:172: ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...)::'lambda'()::operator()() const @ 0xf133acc in /usr/bin/clickhouse
2020.09.14 13:41:08.843563 [ 9598 ] {} <Fatal> BaseDaemon: 13. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:1714: std::__1::__function::__func<ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...)::'lambda'(), std::__1::allocator<ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...)::'lambda'()>, void ()>::operator()() @ 0xf1338fd in /usr/bin/clickhouse
2020.09.14 13:41:08.844069 [ 9598 ] {} <Fatal> BaseDaemon: 14. /build/obj-x86_64-linux-gnu/../src/Common/CurrentMetrics.h:74: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xf1202ea in /usr/bin/clickhouse
2020.09.14 13:41:08.853371 [ 9598 ] {} <Fatal> BaseDaemon: 15. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:2615: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()> >(void*) @ 0xf12d11c in /usr/bin/clickhouse
2020.09.14 13:41:08.854575 [ 9598 ] {} <Fatal> BaseDaemon: 16. start_thread @ 0x9669 in /usr/lib/x86_64-linux-gnu/libpthread-2.30.so
2020.09.14 13:41:08.854730 [ 9598 ] {} <Fatal> BaseDaemon: 17. clone @ 0x1222b3 in /usr/lib/x86_64-linux-gnu/libc-2.30.so
2020.09.14 13:41:08.918098 [ 9598 ] {} <Information> SentryWriter: Not sending crash report
```

**Comment by alexey-milovidov** [comment]
Not a bugfix.

**Pull request merged**